sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriterV2.scala

```diff
@@ -343,3 +402,312 @@ trait CreateTableWriter[T] extends WriteConfigMethods[CreateTableWriter[T]] {
+   */
+  def delete(): DataFrameWriterV2[T] = {
+    dfWriter.notMatchedActions = dfWriter.notMatchedActions :+ DeleteAction(condition)
+    this.dfWriter
+  }
+}
+
+/**
+ * A class for defining actions to be performed when there is no match by source
+ * during a merge operation in a DataFrameWriterV2.
+ *
+ * @param dfWriter the DataFrameWriterV2 instance to which the merge actions will be applied.
+ * @param condition an optional condition to be used with the merge actions.
+ * @tparam T the type parameter for the DataFrameWriterV2.
+ */
+case class WhenNotMatchedBySource[T] (
+    dfWriter: DataFrameWriterV2[T],
+    condition: Option[Expression]) {
+
+  /**
+   * Specifies an action to update all non-matched rows in the target DataFrame when
+   * not matched by the source.
+   *
+   * @return The DataFrameWriterV2 instance with the update all action configured.
+   */
+  def updateAll(): DataFrameWriterV2[T] = {
+    dfWriter.notMatchedBySourceActions =
+      dfWriter.notMatchedBySourceActions :+ UpdateStarAction(condition)
+    this.dfWriter
+  }
+
+  /**
+   * Specifies an action to update non-matched rows in the target DataFrame with the provided
+   * column assignments when not matched by the source.
+   *
+   * @param set A Map of column names to Column expressions representing the updates to be applied.
+   * @return The DataFrameWriterV2 instance with the update action configured.
+   */
+  def update(set: Map[String, Column]): DataFrameWriterV2[T] = {
+    dfWriter.notMatchedBySourceActions = dfWriter.notMatchedBySourceActions :+
+      UpdateAction(condition, set.map(x => Assignment(expr(x._1).expr, x._2.expr)).toSeq)
+    this.dfWriter
+  }
+
+  /**
+   * Specifies an action to insert all non-matched rows into the target DataFrame when not
+   * matched by the source.
+   *
+   * @return The DataFrameWriterV2 instance with the insert all action configured.
+   */
+  def insertAll(): DataFrameWriterV2[T] = {
+    dfWriter.notMatchedBySourceActions =
+      dfWriter.notMatchedBySourceActions :+ InsertStarAction(condition)
+    this.dfWriter
+  }
+
+  /**
+   * Specifies an action to insert non-matched rows into the target DataFrame with the provided
+   * column assignments when not matched by the source.
+   *
+   * @param set A Map of column names to Column expressions representing the values to be inserted.
+   * @return The DataFrameWriterV2 instance with the insert action configured.
+   */
+  def insert(set: Map[String, Column]): DataFrameWriterV2[T] = {
+    dfWriter.notMatchedBySourceActions = dfWriter.notMatchedBySourceActions :+
+      InsertAction(condition, set.map(x => Assignment(expr(x._1).expr, x._2.expr)).toSeq)
+    this.dfWriter
+  }
+
+  /**
+   * Specifies an action to delete non-matched rows from the target DataFrame when not matched by
+   * the source.
+   *
+   * @return The DataFrameWriterV2 instance with the delete action configured.
+   */
+  def delete(): DataFrameWriterV2[T] = {
```

Comment:
ditto, the SQL grammar does not have DELETE clause in `WHEN NOT MATCHED BY SOURCE`

Reply:
Insert was removed